Library/Homebrew/utils/pypi.rb

```diff
@@ -0,0 +1,116 @@
+  def get_pypi_info(package, version)
+    metadata_url = "https://pypi.org/pypi/#{package}/#{version}/json"
+    out, _, status = curl_output metadata_url, "--location"
+
+    return unless status.success?
+
+    begin
+      json = JSON.parse out
+    rescue JSON::ParserError
+      return
+    end
+
+    sdist = json["urls"].find { |url| url["packagetype"] == "sdist" }
+    [json["info"]["name"], sdist["url"], sdist["digests"]["sha256"]]
+  end
+
+  def update_python_resources!(formula, version = nil, print_only: false, silent: false,
+                               ignore_non_pypi_packages: false)
+    # PyPI package name isn't always the same as the formula name. Try to infer from the URL.
+    pypi_name = if formula.stable.url.start_with?("https://files.pythonhosted.org/packages/")
+      File.basename(formula.stable.url).match(/^(.+)-[a-z\d.]+$/)[1]
+    else
+      formula.name
+    end
+
+    version ||= formula.version
+
+    if get_pypi_info(pypi_name, version).blank?
+      odie "\"#{pypi_name}\" at version #{version} is not available on PyPI." unless ignore_non_pypi_packages
+      return
+    end
+
+    non_pypi_resources = formula.resources.reject do |resource|
+      resource.url.start_with? "https://files.pythonhosted.org/packages/"
+    end
+
+    if non_pypi_resources.present? && !print_only
+      odie "\"#{formula.name}\" contains non-PyPI resources. Please update the resources manually."
+    end
+
+    ohai "Retrieving PyPI dependencies for \"#{pypi_name}==#{version}\"" if !print_only && !silent
+    pipgrip_output = Utils.popen_read Formula["pipgrip"].bin/"pipgrip", "--json", "#{pypi_name}==#{version}"
+    unless $CHILD_STATUS.success?
+      odie <<~EOS
+        Unable to determine dependencies for \"#{pypi_name}\" because of a failure when running
+        `pipgrip --json #{pypi_name}==#{version}`. Please update the resources for \"#{formula.name}\" manually.
+      EOS
+    end
+
+    packages = JSON.parse(pipgrip_output).sort.to_h
+
+    # Remove extra packages that may be included in pipgrip output
+    exclude_list = %W[#{pypi_name} argparse pip setuptools wheel wsgiref]
+    packages.delete_if do |package|
+      exclude_list.include? package
+    end
+
+    new_resource_blocks = ""
+    packages.each do |package, package_version|
+      name, url, checksum = get_pypi_info package, package_version
+      # Fail if unable to find name, url or checksum for any resource
+      if name.blank?
+        odie "Unable to resolve some dependencies. Please update the resources for \"#{formula.name}\" manually."
+      elsif url.blank? || checksum.blank?
+        odie <<~EOS
+          Unable to find the URL and/or sha256 for the \"#{name}\" resource.
+          Please update the resources for \"#{formula.name}\" manually.
+        EOS
+      end
+
+      # Indent and append each resource block
+      new_resource_blocks += "  " + <<~EOS.chomp.gsub("\n", "\n  ") + "\n\n"
```

Comment:
`.chomp.gsub("\n", "\n  ") + "\n\n"` seems a bit weird; can that be added to the `EOS` block instead? Also, for indentation, consider not using `<<~` but just doing it yourself.

Reply:
Yeah, it's not super pretty.

> `.chomp.gsub("\n", "\n ") + "\n\n"` seems a bit weird; can that be added to the `EOS` block instead?

It would have to be:

```ruby
    packages.each do |package, package_version|
      ...
      # Indent and append each resource block
      new_resource_blocks += <<-EOS
  resource "#{name}" do
    url "#{url}"
    sha256 "#{checksum}"
  end

      EOS
    end
```

Which I don't think if so great either.

> Also, for indentation, consider not using `<<~` but just doing it yourself.

I'm not sure what you mean. Either of these options?

```ruby
    packages.each do |package, package_version|
      ...
      # Indent and append each resource block
      # Either
      new_resource_blocks += "  resource \"#{name}\" do\n    url \"#{url}\"\n    sha256 \"#{checksum}\"\n  end\n\n"
      # Or
      new_resource_blocks += "  resource \"#{name}\" do\n"
      new_resource_blocks += "    url \"#{url}\"\n"
      new_resource_blocks += "    sha256 \"#{checksum}\"\n"
      new_resource_blocks += "  end\n\n"
    end
```

Reply:
> Which I don't think if so great either.

That seems alright to me!



> I'm not sure what you mean. Either of these options?

Sorry, I just meant what you had above with `<<-EOS`.

Reply:
Fair enough. I can live with it.